packages/python/plotly/plotly/express/_core.py

```diff
@@ -1007,6 +1007,99 @@ def build_dataframe(args, attrables, array_attrables):
     return args
 
 
+def _discrete_agg(x):
+    if len(x) == 1:
+        return x.iloc[0]
+    else:
+        return ""
+
+
+def process_dataframe_hierarchy(args):
+    """
+    Build dataframe for sunburst or treemap when the path argument is provided.
+    """
+    df = args["data_frame"]
+    path = args["path"][::-1]
+
+    # ------------ Define aggregation functions --------------------------------
+    lambda_discrete = _discrete_agg
+    agg_f = {}
+    aggfunc_color = None
+    if args["values"]:
+        try:
+            df[args["values"]] = pd.to_numeric(df[args["values"]])
+        except ValueError:
+            raise ValueError(
+                "Column `%s` of `df` could not be converted to a numerical data type."
+                % args["values"]
+            )
+
+        if args["color"]:
+            if args["color"] == args["values"]:
+                aggfunc_color = "sum"
+        count_colname = args["values"]
+    else:
+        if args["color"]:  # color passed but not value
+            # we need a count column for the weighted mean of color
+            # trick to be sure the col name is unused: take the sum of existing names
+            count_colname = "".join([str(el) for el in list(df.columns)])
+            # we can modify df because it's a copy of the px argument
+            df[count_colname] = 1
+
+    if args["color"]:
+        if df[args["color"]].dtype.kind not in "bifc":
+            aggfunc_color = lambda_discrete
+        elif not aggfunc_color:
+            aggfunc_color = lambda x: np.average(
+                x, weights=df.loc[x.index, count_colname]
+            )
+        agg_f[args["color"]] = aggfunc_color
+    if args["color"] or args["values"]:
+        agg_f[count_colname] = "sum"
+
+    #  Other columns (for color, hover_data, custom_data etc.)
+    cols = list(set(df.columns).difference(path))
+    for col in cols:  # for hover_data, custom_data etc.
+        if col not in agg_f:
+            agg_f[col] = lambda_discrete
+    # ----------------------------------------------------------------------------
+
+    df_all_trees = pd.DataFrame(columns=["labels", "parent", "id"] + cols)
+    #  Set column type here (useful for continuous vs discrete colorscale)
+    for col in cols:
+        df_all_trees[col] = df_all_trees[col].astype(df[col].dtype)
+    for i, level in enumerate(path):
+        df_tree = pd.DataFrame(columns=df_all_trees.columns)
+        if not agg_f:
+            dfg = df.groupby(path[i:]).sum(numerical_only=True)
```

Comment:
under what circumstances do we do this, and what's the reasoning?

Reply:
it's if neither `values` nor `color` was passed. I think I need to use some sort of aggregation function because after I call `reset_index` to get the groups, but maybe it's possible to do it in a more elegant/efficient way.

Reply:
I could get rid of this part with the `count_colname` logic.